### ISSUE TITLE
Create tooling to backup/wipe radio partition

### ISSUE THREAD

**Issue by patcon** [patcon/mission-impossible-android#33]
@mikeperry-tor mentioned in IRC.

Links:
- https://blog.torproject.org/blog/mission-impossible-hardening-android-security-and-privacy#comment-55280
- http://rootzwiki.com/topic/25016-modany-phone-any-rom-wi-fi-only-mode-disable-cell-radio/

Did a little investigation, and if we wanted to make this simple for any platform, would probably steal some things out of the [`repo breakfast <device_codename>`](http://wiki.cyanogenmod.org/w/Envsetup_help#breakfast) command to fetch the right device repo on github, and then parse the `fstab.<device_codename>` and grep for the line mentioning "radio".

Example:
https://github.com/CyanogenMod/android_device_lge_mako/blob/cm-12.0/fstab.mako#L10

**Comment by n8fr8** [comment]
+1 +1 +1 definitely important!

**Comment by patcon** [comment]
So I'm realizing one possible benefit to this radio wipe approach rather than the easier "buy a wifi tablet" approach: presumably Android on a device with a wiped radio will still act like SMS/MMS is enabled, even after you've wiped the radio partition. Conversely, CM for nexus 7 lacks lots of SMS-related apps and defaults that any SMS app expects to be present.

Why is this helpful? I'm currently working to make Voice+ more adaptable. VoicePlus currently only works with Google Voice, intercepting SMS's and letting normal sms apps send/receive data-based SMS backed by Google Voice. I'm hoping to get Voice+ working with an arbitrary data-based SMS service. Why? Because while there's a hope of finding one of those that doesn't use GCM, there is no hope in hell of GVoice ever shipping without a GCM dependency.

So basically, all of this is a step toward having a no-GApps phone that can send/receive texts via data as if there were regular SMS, using regular SMS apps like TextSecure :)

Why do all this? Because this allows the use of any standard SMS apps to send Google Voice data-channel texts that go to regular phone users. In particular, we getting this to work allows textsecure, that can communicate with normal cellular SMS users.

**Comment by patcon** [comment]
https://github.com/android/platform_hardware_ril/blob/master/rild/radiooptions.c#L28
http://forum.xda-developers.com/showthread.php?t=1505643

Hopefully that binary and init will keep the battery drain from happening, but I'll have to investigate

**Comment by patcon** [comment]
```
# Backup
adb shell
su
busybox dd if=/dev/block/platform/msm_sdcc.1/by-name/modem of=/sdcard/modem.img.bak bs=1M
exit
adb pull /sdcard/modem.img.bak local/path/to/

# Wiping radio
adb reboot-recovery # Doesn't work when running normally
adb shell
busybox dd if=/dev/zero of=/dev/block/platform/msm_sdcc.1/by-name/modem bs=1M

# Re-flashing with backup
adb reboot-bootloader
fastboot flash modem local/path/to/modem.img.bak

```

Might be able to skip the need to adb in and zero the partition if we create a blank `zero.img`

**Comment by SchnWalter** [comment]
I think it is about that that someone creates a python wrapper around adb, what do you think a new repo? or just a .py in the utils folder for now so that we don't waste to many resources on it.

**Comment by patcon** [comment]
Maybe pop open another issue for that?

**Comment by patcon** [comment]
Might help me `adb shell` in and see why it's hanging:
https://groups.google.com/forum/#!topic/mozilla.dev.b2g/epQ6qhIFZ50

**Comment by patcon** [comment]
Liveboot didn't really help, as it didn't kick in until after the FDE password part, which my phone didn't get to.

Since I hadn't actually done all the `pm disable <foo>` stuff form [this comment](https://blog.torproject.org/blog/mission-impossible-hardening-android-security-and-privacy#comment-55280), I went back and did that before doing `fastboot erase modem` (presumably the same). Still no dice, but got the `/proc/last_kmsg` log this time:

https://gist.github.com/patcon/2d61ea506a142307b31b

I'll investigate it later

**Comment by mikeperry-tor** [comment]
I noticed two things in the kmsg log:
1. As you noted in your edit, it seems to be complaining about trying to mount the modem partition. Maybe simply making it a formatted but empty (ext2fs?) partition might work?
2. The kernel is being given androidboot.baseband=mdm in the boot params. Might be something to look into removing if simply making the partition mountable or removing it from fstab is not enough.

I can look into this later with my moto e, but I'm rather swamped at the moment.

**Comment by patcon** [comment]
Did some experimentation around this blog comment: https://blog.torproject.org/blog/mission-impossible-hardening-android-security-and-privacy#comment-55280

```
adb shell
su
pm disable com.android.cellbroadcastreceiver
pm disable com.android.mms
pm disable com.android.providers.telephony
pm disable com.android.smspush
pm disable com.cyanogenmod.samsungservicemode # Not on my phone

# Special consideration
pm disable com.android.phone
```

Ideally, I (and we?) still want 1) SIP calling with the native dialler to work, and 2) VoicePlus SMS message intercept to work.


#### Native SIP dialler

`com.android.phone` needs to be enabled or else trying to view the dialler settings (where SIP is configured), crashes the dialler. Native SIP calls (incoming and outgoing) only work went that's enabled. 

#### VoicePlus SMS interception

As for VoicePlus working (ie. TextSecure properly intercepting incoming GVoice SMS), that doesn't seem to depend on any of those packages being enabled, oddly. Even setting `adb shell "su root radiooptions 1"` still allows SMS interception to happen. Having said that, VoicePlus has some checks during account setup and Gvoice token retrieval, and that process was done before disabling packages. I still probably need to sort out how disabling packages affects the full VoicePlus auth process, so that I can make sure it continues to work for apps like VoicePlus and [TextNowPlus](https://github.com/patcon/textnowplus-android).

**Comment by mikeperry-tor** [comment]
I got around to trying to wipe the modem firmware on my moto e. Reformatting the /dev/block/platform/msm_sdcc.1/by-name/modem partition as ext2 allowed me to boot, but the systemui process kept crashing. So I dug a little further, and discovered that my modem device was mounted on /firmware. If I remounted that read-write, I could remove just the modem files from /firmware/images/ and this seemed to fully disable the cell radio (Settings->About Phone says my baseband is unknown, I have no signal, and unknown IMEI).

Unfortunately, this also completely disables GPS. That is probably useful for some people, but I personally wish it was independent.

```
adb shell
su
pm disable com.android.cellbroadcastreceiver
pm disable com.android.mms
pm disable com.android.providers.telephony
pm disable com.android.smspush

mount -o remount,rw /firmware
rm /firmware/image/modem.*
mount -o remount,ro /firmware
reboot
```
